sympy/polys/agca/extensions.py

```diff
@@ -67,7 +73,14 @@ def __rsub__(f, g):
     def __mul__(f, g):
         rep = f._get_rep(g)
         if rep is not None:
-            return ExtElem((f.rep*rep) % f.ext.mod, f.ext)
+            return ExtElem((f.rep * rep) % f.ext.mod, f.ext)
+        else:
+            return NotImplemented
+
+    def __mod__(f, g):
+        rep = f._get_rep(g)
+        if rep is not None:
+            return ExtElem((f.rep % rep) % f.ext.mod, f.ext)
```

Comment:
I don't think that `__mod__` is well defined in this class. `f.rep` is usually the canonical representative of lowest degree but the formula should work for all representatives. So `((f.rep + f.ext.mod) % rep)  % f.ext.mod` should be the same as `(f.rep % rep) % f.ext.mod`. This is generally true only if `f.ext.mod % rep` is zero, or `rep` is a factor of `f.ext.mod`.

Will this be needed somewhere?

Reply:
It is needed when constructing a nested extension. If I remove this method then run this:
```python
from sympy import *
from sympy.polys.agca.extensions import FiniteExtension
from sympy.abc import x, y

Kx = FiniteExtension(Poly(x**2 - 2, x, domain=ZZ))
Ky = FiniteExtension(Poly(y**2 - 2, y, domain=ZZ))
Kxy = FiniteExtension(Poly(y**2 - 2, y, domain=Kx))
```
That gives
```console
$ python t.py 
Traceback (most recent call last):
  File "t.py", line 7, in <module>
    Kxy = FiniteExtension(Poly(y**2 - 2, y, domain=Kx))
  File "/Users/enojb/current/sympy/sympy/sympy/polys/agca/extensions.py", line 222, in __init__
    mod, rem = mod.div(mod.LC())
  File "/Users/enojb/current/sympy/sympy/sympy/polys/polytools.py", line 1680, in div
    q, r = F.div(G)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/polyclasses.py", line 491, in div
    q, r = dmp_div(F, G, lev, dom)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/densearith.py", line 1629, in dmp_div
    return dmp_rr_div(f, g, u, K)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/densearith.py", line 1372, in dmp_rr_div
    return dup_rr_div(f, g, K)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/densearith.py", line 1337, in dup_rr_div
    if lc_r % lc_g:
TypeError: unsupported operand type(s) for %: 'ExtensionElement' and 'ExtensionElement'
```
Maybe there's a better way.

Reply:
Perhaps mod should only be defined for constants.

Reply:
It looks like this could work for constants.

Reply:
Invertible constants, and the result would be zero.

Reply:
I've changed it so that `f % g` raises `NotImplementedError` unless `g.is_ground`.